[WIP] Added python api for downloading traefik/etcd

GeorgianaElena opened the pull request:
Besides downloading traefik, I also added a function to download etcd.
I checked out TLJH where something similar is done and I saw the download is checked against the public sha256 checksum. However, I didn't find any for the latest traefik binary on the releases page. Should I search it somewhere else?

GeorgianaElena commented:
Also, should I take care here of the downloaded binaries to be on PATH?

minrk commented on jupyterhub_traefik_proxy/install.py:
```diff
@@ -0,0 +1,74 @@
+
+
+def main():
+    deps_dir = os.path.join(HERE, DEPS_INSTALL_DIR_NAME)
```
I would add an `-o, --out` argument to store the destination, and use the current directory by default.

minrk commented:
For the checksums, you can calculate it yourself by downloading the file and running the checksum if they haven't published it themselves.

GeorgianaElena commented:
@minrk , thanks for the feedback. I made the installer store only the binaries(traefik, etcd, etcdctl) at the passed output directory. 
In travis I tried to directly download the binaries in /usr/local/bin so that they will directly be on PATH. But I need sudo for accessing /usr/local/bin. Is it wrong if in the installer I run comands with sudo whenever I get a PermissionError?

minrk commented on .travis.yml:
```diff
@@ -16,13 +16,7 @@ before_install:
-  - tar xzvf etcd-v3.3.10-linux-amd64.tar.gz
-  - sudo cp ./etcd-v3.3.10-linux-amd64/etcd /usr/local/bin
-  - sudo cp ./etcd-v3.3.10-linux-amd64/etcdctl /usr/local/bin
+  - sudo python -m jupyterhub_traefik_proxy.install --output=/usr/local/bin
```
Try:

    sudo -E python ...

to avoid `python` resolving to a different version

minrk replied:
...or `sudo $(which python) -m ...`

minrk commented:
> Is it wrong if in the installer I run commands with sudo whenever I get a PermissionError?

Not necessarily, but sudo sometimes comes with confusion of paths. The alternative is to add a directory to $PATH that you have write permission to:

```
python -m traefik.install --output ./bin
export PATH=$PWD/bin:$PATH
```

GeorgianaElena commented:
closes https://github.com/jupyterhub/traefik-proxy/issues/2

minrk commented on jupyterhub_traefik_proxy/install.py:
```diff
@@ -0,0 +1,227 @@
+
+    if os.path.exists(traefik_bin):
+        print(f"Traefik already exists")
+        checksum = checksum_file(traefik_bin)
```
Instead of checksumming the binary, let's checksum the file we download. Then it's just one per URL (the keys can be the URLs we get it from). If folks start publishing checksums, it will be for the archive, not individual files.

minrk commented on jupyterhub_traefik_proxy/install.py:
```diff
@@ -0,0 +1,227 @@
+            os.chmod(traefik_bin, 0o755)
+            print("--- Done ---")
+            return
+        else:
```
Let's warn instead of error if they requests a version/platform that we don't have a checksum for

GeorgianaElena commented:
@minrk I addressed your comments above.
traefik and etcd are installed even if we don't have checksums for those specific versions and we warn the user about this. Please let me know if you spot anything else I should change.
